File: code/dobot_quad_sdk-main/high_level/python/examples/e3_auto_state_switch.py
#!/usr/bin/env python3
"""Example 3: Switch robot to a target state (interactive or CLI).
Usage: python e3_auto_state_switch.py [server_address] [state_name]
"""
import sys
from dobot_quad import RobotClient

STATES = [
    "emergency",
    "ready",
    "stand_down",
    "balance_stand",
    "walk",
    "rl",
    "flying_trot",
    "wave",
    "dance0",
    "jump",
    "recovery",
    "change_mode",
]

WHEEL_STATES = [
    "emergency",
    "ready",
    "stand_down",
    "wheel_loco",
    "drift",
    "handstand",
    "change_mode",
]


def main():
    robot = RobotClient(sys.argv[1] if len(sys.argv) > 1 else "192.168.5.2:50051")
    robot.enable_safety_ready()

    states = WHEEL_STATES if robot.is_quad_wheel() else STATES

    if len(sys.argv) > 2:
        target = sys.argv[2].lower()
    else:
        print("Available states:")
        for i, s in enumerate(states):
            print(f"  {i}. {s}")

        while True:
            try:
                choice = int(input(f"Select [0-{len(states)-1}]: "))
                if 0 <= choice < len(states):
                    break
                print(f"Invalid choice. Please select [0-{len(states)-1}].")
            except ValueError:
                print("Invalid input. Please enter a number.")
            except (KeyboardInterrupt, EOFError):
                print("\nAborted.")
                return 0

        target = states[choice]

    print(f"Switching to: {target}")
    if target == "change_mode":
        res = robot.change_mode()
    else:
        res = robot.set_target_state(target)

    if res and res.success:
        return 0
    print(f"Failed: {res.message if res else 'cancelled'}")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: code/dobot_quad_sdk-main/high_level/cpp/e3_auto_state_switch.cpp
// Example 3: Switch robot to a target state (interactive or CLI).
// Usage: ./e3_auto_state_switch [server_address] [state_name]
//
// Simplified API:
//   client.set_balance_stand();          // direct call
//   client.set_target_state("walk");     // by name

#include "robot_client.h"
#include <algorithm>
#include <limits>

int main(int argc, char** argv)
{
    robot::Client client(argc > 1 ? argv[1] : "192.168.5.2:50051");
    robot::enable_safety_ready(client);

    auto run_action = [&client](const std::string& target) {
        if (target == "change_mode") {
            return client.change_mode();
        }
        return client.set_target_state(target);
    };

    if (argc > 2) {
        // CLI mode: ./e3 addr balance_stand
        std::string target = argv[2];
        std::transform(target.begin(), target.end(), target.begin(), ::tolower);
        return run_action(target) ? 0 : 1;
    }

    // Interactive mode
    const bool is_wheel = client.is_quad_wheel();

    const char* quad_states[] = {"emergency", "ready", "stand_down", "balance_stand", "walk", "rl", "flying_trot",
        "wave", "dance0", "jump", "recovery", "change_mode"};
    const char* wheel_states[] = {"emergency", "ready", "stand_down", "wheel_loco", "drift", "handstand",
        "change_mode"};

    const char** states = is_wheel ? wheel_states : quad_states;
    const int n = is_wheel ? static_cast<int>(sizeof(wheel_states) / sizeof(wheel_states[0]))
                           : static_cast<int>(sizeof(quad_states) / sizeof(quad_states[0]));

    std::cout << "Available states:" << std::endl;
    for (int i = 0; i < n; ++i)
        std::cout << "  " << i << ". " << states[i] << std::endl;

    int choice = -1;
    while (true) {
        std::cout << "Select [0-" << n - 1 << "]: ";
        if (!(std::cin >> choice)) {
            if (std::cin.eof())
                return 0;
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cerr << "Invalid input. Please enter a number." << std::endl;
            continue;
        }
        if (choice < 0 || choice >= n) {
            std::cerr << "Invalid choice " << choice << ". Please select [0-" << n - 1 << "]." << std::endl;
            continue;
        }
        break;
    }

    std::string target = states[choice];
    std::cout << "Switching to: " << target << std::endl;

    return run_action(target) ? 0 : 1;
}
